### ISSUE TITLE
Ability to specify columns in CSV which I want to have in DataFrame

### ISSUE THREAD

**Issue by savinov** [databricks/spark-csv#321]
I wounder if I can specify column numbers which I want to see in DataFrame. Also I want to specify shema for this columns.
For example I have a CSV with columns C0..C5. I want to have C0, C2 and C5 in DataFrame and I want to specify that C0 is a String, C2 is a Number and C5 is a date.

Current implementation load all columns from the specified schema from the column C0 to the end of schema (StructType fields).
For example: I have a CSV with columns C0..C5 and I specify schema with 3 fields: String, Number, Date.
In result DataFrame I will have C0 with type String, C1 with type Number and C2 with type Date.

Please consider this feature, I think it's usefull.

**Comment by HyukjinKwon** [comment]
Actually, this library supports 'PrunedFilterScan` meaning it reads, from the disk, the CSV file all but does not try to convert not required fields. This was hanlded in https://github.com/databricks/spark-csv/pull/169.

Please use `select` the columns you want. This will try to not parse not selected columns.

**Comment by savinov** [comment]
Ok, it's performance part of my request and probably resolved by #169, but I still have to specify all the columns in the schema that CSV contains.
I think it's a good idea to add an option with the list of column indexes or a map that maps column names to column indexes. This will make possible to don't care about all the column types in CSV but only columns that I interested in.

**Comment by HyukjinKwon** [comment]
Hm.. For me I am not too sure if that option would be needed. There are workarounds for this. I haven't tested them yet but should work.

- **First** : Create RDD with this library, drop some columns with manual transformations, and then load the RDD to DataFrame with this library.

    ```scala
val indexToDrop = 1
val path = "path-to-file"

val csvRDD = sparkContext.textFile(path)
val droppedCsvRDD = csvRdd.map(_.split(",")).map(_.drop(indexToDrop))
val df = new CsvParser()
  .withUseHeader(true)
  .csvRdd(sqlContext, droppedCsvRDD)
  .collect()
```

- **Second** : Do not infer schema with this library, select columns you want and then give types.

    ```scala
val path = "path-to-file"

val df = sqlContext.read
  .format("csv")
  .option("inferSchema", "false")
  .load(path)
  .select("field-to-select")

val prunedDf = df.withColumn("double-field-to-select", toDouble(df("field-to-select"")))
```

**Comment by HyukjinKwon** [comment]
@savinov Would that make sense if we perform some transformations with columns within DataFrame or RDD APIs?

**Comment by savinov** [comment]
Oh, I forget transformations.
Yeah, I think it might be done this way: transform RDD by dropping unneded columns.

I'm not sure about casting that it's needed: if all the unneded columns are dropped already we can infer schema, right?

**Comment by HyukjinKwon** [comment]
@savinov indeed. I got your point of this issue but maybe this might be better closed for now. Definitely this should be reopened if the workarounds above is found heavy performance regressions!

**Comment by savinov** [comment]
This works:
1. load CSV with spark-csv
2. select needed columns
3. save it to HDFS
4. load with spark-csv file with dropped columns
5. provide schema for all columns in this file

It seems as a workaround, but not ugly.
If spark CSV simplify this steps, it'll be a good feature.

**Comment by chumo** [comment]
I still think that @savinov proposed feature would make a lot of sense. Consider the case in which your csv files contain 100 columns (with no header), and you are interested in loading only columns C0, C12 and C50.

**Comment by irfanaziz098** [comment]
Let me share a small finding;
Lets say you have a  csv file with columns c1,c2,c3,c4,c5
Now if you want to load only c1 cand c2 this can be done by specifiying only two column names in schema.
```
val imsi_refSchema = StructType(Array(
    StructField("c1", StringType, true),
    StructField("c2", StringType, true)
  ))
```
However, if you want only c1 and c3 i.e not consecutive  columns, this is not possible.
```
val imsi_refSchema = StructType(Array(
    StructField("c1", StringType, true),
    StructField("c3", StringType, true)
  ))
```
The above schema will have the values on c1 in c1 and the values of c2 will be loaded as c3.

Hope this helps
